src/irq.c

```diff
@@ -68,153 +68,167 @@ static int irq_config(const char *key, const char *value) {
+  return (int)fields_num;
+}
 
+static int irq_read_data(metric_family_t *fam) {
   /*
    * Example content:
    *         CPU0       CPU1       CPU2       CPU3
    * 0:       2574          1          3          2   IO-APIC-edge      timer
    * 1:     102553     158669     218062      70587   IO-APIC-edge      i8042
    * 8:          0          0          0          1   IO-APIC-edge      rtc0
    */
-  fh = fopen("/proc/interrupts", "r");
+  FILE *fh = fopen("/proc/interrupts", "r");
   if (fh == NULL) {
     ERROR("irq plugin: fopen (/proc/interrupts): %s", STRERRNO);
     return -1;
   }
 
   /* Get CPU count from the first line */
-  if (fgets(buffer, sizeof(buffer), fh) != NULL) {
-    cpu_count = strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
-  } else {
-    ERROR("irq plugin: unable to get CPU count from first line "
-          "of /proc/interrupts");
+  char cpu_buffer[1024] = {0};
+  if (fgets(cpu_buffer, sizeof(cpu_buffer), fh) == NULL) {
+    ERROR("irq plugin: unable to get CPU count from first line of "
+          "/proc/interrupts");
     fclose(fh);
-    return -1;
+    return EINVAL;
   }
 
+  char *cpu_names[256] = {0};
+  int cpu_count = strsplit(cpu_buffer, cpu_names, STATIC_ARRAY_SIZE(cpu_names));
+  for (int i = 0; i < cpu_count; i++) {
+    if (strncmp(cpu_names[i], "CPU", 3) == 0) {
+      cpu_names[i] += 3;
+    }
+  }
+
+  char buffer[1024] = {0};
   while (fgets(buffer, sizeof(buffer), fh) != NULL) {
-    char *irq_name;
-    size_t irq_name_len;
-    derive_t irq_value;
-    int i;
-    int fields_num;
-    int irq_values_to_parse;
-
-    fields_num = strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
+    strstripnewline(buffer);
+
+    /* There is one column with the interrupt ID before the CPU counters and
+     * there may be up to three columns after the counters. */
+    char *fields[cpu_count + 8];
+    memset(fields, 0, sizeof(fields));
+
+    int fields_num = irq_strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
     if (fields_num < 2)
       continue;
 
     /* Parse this many numeric fields, skip the rest
      * (+1 because first there is a name of irq in each line) */
-    if (fields_num >= cpu_count + 1)
-      irq_values_to_parse = cpu_count;
-    else
+    int irq_values_to_parse = cpu_count;
+    if (fields_num < cpu_count + 1) {
       irq_values_to_parse = fields_num - 1;
+    }
 
     /* First field is irq name and colon */
-    irq_name = fields[0];
-    irq_name_len = strlen(irq_name);
-    if (irq_name_len < 2)
+    char *irq_name = fields[0];
+    size_t irq_name_len = strlen(irq_name);
+    if ((irq_name_len < 2) || (irq_name[irq_name_len - 1] != ':')) {
       continue;
+    }
+    /* strip colon */
+    irq_name[irq_name_len - 1] = 0;
+    irq_name_len--;
 
-    /* Check if irq name ends with colon.
-     * Otherwise it's a header. */
-    if (irq_name[irq_name_len - 1] != ':')
+    /* Is it the the ARM fast interrupt (FIQ)? */
+    if (strcmp(irq_name, "FIQ") == 0)
```

Comment:
IMHO this could be in user config, but that could be regression to old users.

Reply:
Yeah, I don't think this belongs here, but I don't fully understand why the code is there yet and so didn't want to remove it in the middle of everything else.

Reply:
FIQ is just other type of interrupt:
* https://en.wikipedia.org/wiki/Fast_interrupt_request
* https://stackoverflow.com/questions/973933/what-is-the-difference-between-fiq-and-irq-interrupt-system

That code was added for: https://github.com/collectd/collectd/issues/971

But IMHO that was wrong fix.  It should check for the error condition, not IRQ name.

Reply:
Okay, I'm pretty sure that the current code does not try to parse `NULL` as a value, and if it does we can fix that code instead. I'll revert that change.

There are a few rows in Linux that have a single column, e.g. `ERR`. They are currently attributed to "CPU 0", which is probably not 100% correct. I suspect FIQ was a similar case.